src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,168 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final byte INITIAL_CAPACITY = 10;
+    private static final int MAX_CAPACITY = Integer.MAX_VALUE;
+    private static final double CAPACITY_MULTIPLIER = 1.5;
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[INITIAL_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
+        if (size == elementData.length) {
+            increaseCapacity();
+        }
 
+        insertElementAt(size, value);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " is out of bounds");
+        }
+
+        if (size == elementData.length) {
+            increaseCapacity();
+        }
 
+        if (index < size) {
+            shiftElementsRightAt(index);
+        }
+
+        insertElementAt(index, value);
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null) {
+            throw new RuntimeException("List must not be a null");
+        }
+
+        if (size + list.size() > elementData.length) {
+            increaseCapacity();
+        }
 
+        for (int i = 0, j = size; i < list.size(); i++, j++) {
+            insertElementAt(j, list.get(i));
+        }
```

Comment:
Here you can do much easier. It is enough just to go through each element and use the ready-made add(value) method